nupic/research/spatial_pooler.py

```diff
@@ -997,59 +959,97 @@ def _raisePermanenceToThreshold(self, perm, mask):
     @param perm:    An array of permanence values for a column. The array is
                     "dense", i.e. it contains an entry for each input bit, even
                     if the permanence value is 0.
-    @param mask:    the indices of the columns whose permanences need to be
-                    raised.
     """
-    if len(mask) < self._stimulusThreshold:
+    synapses = self._layer.columns[index].segment.synapses
+    if len(synapses) < self._stimulusThreshold:
       raise Exception("This is likely due to a " +
       "value of stimulusThreshold that is too large relative " +
-      "to the input size. [len(mask) < self._stimulusThreshold]")
+      "to the input size. [len(synapses) < self._stimulusThreshold]")
 
-    numpy.clip(perm, self._synPermMin, self._synPermMax, out=perm)
+    numConnected = sum(1 for synapse in synapses if synapse.permanence >= self._synPermConnected)
     while True:
-      numConnected = numpy.nonzero(perm > self._synPermConnected)[0].size
       if numConnected >= self._stimulusThreshold:
         return
-      perm[mask] += self._synPermBelowStimulusInc
+      numConnected = 0
```

Comment:
>Why is numConnected being re-calculated here and below, compared to above?

Because according to current implementation, before permanences can be incremented with `self._synPermBelowStimulusInc` constant, first we have to check numConnected fits criteria. We could remove the first counting operation, but this would require we use a loop counter to check `synapse.permanence >= self._synPermConnected` only after 1 iteration:

```
i = 0
while True:
  if numConnected >= self._stimulusThreshold:
    return
  numConnected = 0
  for synapse in synapses:
    if i > 0:
      synapse.permanence += self._synPermBelowStimulusInc
    if synapse.permanence >= self._synPermConnected:
      numConnected += 1
  i += 1
```